### PR TITLE
Coupons: Update analytics for M2

### PR DESCRIPTION
**Pull request by itsmeichigo**
### Description

This PR updates Tracks events for coupon edit and deletion following p91TBi-7vL-p2:
- Track new actions on the coupon details screen: "edit_tapped" and "delete_tapped"
- Track events when coupon deletion succeeds or fails.
- Track events when coupon update is initiated. Here we check to see which value is updated.
- Track events when coupon update succeeds or fails.

### Testing instructions

Build the app and look at the console for the Tracks event in Xcode console:
- Navigate to Menu > Coupons > select a coupon > select the ellipsis button.
- Select Edit Coupon, you should see an event in the console: `🔵 Tracked coupon_details` with one of the properties: `AnyHashable("action"): "tapped_edit"`.
- Edit any value of the coupon and then tap the Save button. You should see an event in the console: `🔵 Tracked coupon_update_initiated`. Look at the properties in that event, and make sure that they are all correct.
- If the update succeeds, you should see `🔵 Tracked coupon_update_success` or `🔵 Tracked coupon_update_failed` otherwise. In the case of failure, the error should also be logged.
- Back to the Coupon Details screen, select the ellipsis button and select "Delete Coupon". The console should log `🔵 Tracked coupon_details` with property `AnyHashable("action"): "tapped_delete"`.
- Proceed to delete the coupon - if the deletion succeeds you should see `🔵 Tracked coupon_delete_success`, otherwise `🔵 Tracked coupon_delete_failed` with an error.

### Screenshots

N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by selanthiraiyan — COMMENTED** [review]
@itsmeichigo 

I can successfully validate most of the test cases. 👏 🎉 

Meanwhile, I noticed the following things while testing. Kindly let me know if I am missing something.

### `coupon_update_initiated` event seems to have incorrect values for some properties.
> Edit any value of the coupon and then tap the Save button. You should see an event in the console: 🔵 Tracked coupon_update_initiated. Look at the properties in that event, and make sure that they are all correct.

I edited ONLY the coupon code, for which I believe that ONLY `coupon_code_updated` should be set as `true`. But, some other properties such as `amount_updated` and `expiry_date_updated` are set as `true` as well. 
```
Tracked coupon_update_initiated, properties: [AnyHashable("amount_updated"): true, AnyHashable("allowed_products_or_categories_updated"): false, AnyHashable("description_updated"): false, AnyHashable("expiry_date_updated"): true, AnyHashable("coupon_code_updated"): true, AnyHashable("is_wpcom_store"): false, AnyHashable("usage_details_updated"): true, AnyHashable("blog_id"): 205007824]
```

### `coupon_delete_failed` event when editing coupon fails.
> If the update succeeds, you should see 🔵 Tracked coupon_update_success or 🔵 Tracked coupon_update_failed otherwise. In the case of failure, the error should also be logged.

I see `coupon_delete_failed` when editing a coupon fails. But, I don't see `coupon_update_failed` event. 

```
Tracked coupon_delete_failed, properties: [AnyHashable("error_code"): "-1001", AnyHashable("is_wpcom_store"): false, AnyHashable("error_domain"): "NSURLErrorDomain", AnyHashable("blog_id"): 205007824, AnyHashable("error_description"): "Error Domain=NSURLErrorDomain Code=-1001 \"The request timed out.\" UserInfo={_kCFStreamErrorCodeKey=-2102, NSUnderlyingError=0x28026a4c0 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 \"(null)\" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <3F9EC138-61F1-4A17-AC4D-AA52800E5346>.<41>, _NSURLErrorRelatedURLSessionTaskErrorKey=(\n    \"LocalDataTask <3F9EC138-61F1-4A17-AC4D-AA52800E5346>.<41>\"\n), NSLocalizedDescription=The request timed out., NSErrorFailingURLStringKey=https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/205007824/rest-api/, NSErrorFailingURLKey=https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/205007824/rest-api/, _kCFStreamErrorDomainKey=4}"]
````

**Comment by itsmeichigo** [comment]
Thank you @selanthiraiyan for the review!

> I edited ONLY the coupon code, for which I believe that ONLY coupon_code_updated should be set as true. But, some other properties such as amount_updated and expiry_date_updated are set as true as well.

I've added changes to fix the issue with incorrect `expiry_date_updated` and `usage_restrictions_updated`. I cannot reproduce the case with incorrect ` amount_updated`, could you test again to see if it still happens?

> I see coupon_delete_failed when editing a coupon fails. But, I don't see coupon_update_failed event.

My bad - I sent the incorrect event name 😅 Updated in 8a03ded1cefe4b1a75ccd323575adf8880949443.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 1]
```diff
@@ -213,6 +213,40 @@ final class AddEditCouponViewModel: ObservableObject {
...
+        }()
+
+        let expiryDateUpdated: Bool = {
+            guard let oldDate = initialCoupon.dateExpires, let newDate = coupon.dateExpires else {
+                return initialCoupon.dateExpires != coupon.dateExpires
+            }
+            return !oldDate.isSameDay(as: newDate)
+        }()
+
+        ServiceLocator.analytics.track(.couponUpdateInitiated, withProperties: [
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Would it make sense to extract the `couponUpdateInitiated` event tracking code to a private function?

**Review by selanthiraiyan — COMMENTED** [review]
About the `amount_updated` issue, I noticed that my site is using comma as a decimal separator. 😅

I hope the following screenshots are helpful. Kindly let me know if you have any questions.

Screenshot 2022-05-11 at 9 55 43 AM

Screenshot 2022-05-11 at 9 57 30 AM

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Good idea! Updated in d19e7999ee73f7044a8a6d187412d8a8d4aacd6e 👍

**Comment by itsmeichigo** [comment]
@selanthiraiyan thanks for clarifying the case with the incorrect amount! There are a few complications in comparing the formatted amounts, so I added a change to compare the actual values in 7ee2dca249a92d6da8851f3965bf57ecbef9424e. This should work properly with different cases including the case when a user updates a value from "0,00" to "000".

This PR is ready for another look, thanks!

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/Coupon Text Validators/CouponAmountInputFormatter.swift [hunk 2]
```diff
@@ -3,17 +3,25 @@ import Foundation
...
-            return PriceInputFormatter().format(input: text)
+            return priceInputFormatter.format(input: text)
         }
         return "0"
     }
 }
+
+extension CouponAmountInputFormatter {
+    func value(from input: String) -> NSNumber? {
+        priceInputFormatter.value(from: input)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/Coupon Text Validators/CouponAmountInputFormatter.swift:29]: Would it make sense to add unit tests for `CouponAmountInputFormatter` and `PriceInputFormatter`?

**Review by selanthiraiyan — APPROVED** [review]
Thanks for making these changes @itsmeichigo. 

The code looks and works great now. 🎉

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/Coupon Text Validators/CouponAmountInputFormatter.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/Coupon Text Validators/CouponAmountInputFormatter.swift:29]: Sure! Added tests in 135ca0af6a6eaf9875493fd07e3dd4b2f986ad74 and 84ae4565341c643e4df047a66728aca83d66d2b2.

**Pull request merged**